### DISCUSSION TITLE
[Share] Show Citation Counts

### DISCUSSION THREAD

**Discussion by volatile-static** [windingwind/zotero-actions-tags#150] [Action Scripts]
### Description

Fetch citation counts from `doi.org` and show in a custom column.
![image](https://github.com/windingwind/zotero-actions-tags/assets/59945270/f3766132-434d-42cd-8dd8-ef2e1d98b64f)
![image](https://github.com/windingwind/zotero-actions-tags/assets/59945270/04bdcc75-7c8e-46ef-82f3-1da342acf6de)

### Event

Program Startup

### Operation

Script

### Data

```javascript
/**
 * Show Citation Counts
 * @author volatile-static
 * @usage plug-and-play
 * @see https://github.com/windingwind/zotero-actions-tags/discussions/150
 */
const pluginID = 'zoterotag@euclpts.com', // 与package.json一致
	window = require('window'),
	_caches = {};  // 缓存获取到的引用数
	
function renderCell(_, data, column) {
	const span = window.document.createElement('span');
	span.classList.add('cell');
	column.className.split(' ').forEach(name => span.classList.add(name));

	if (data.length < 1)  // 没有DOI
		return span;
	if (data[0] == '#' && !_caches[data]) {
		window.fetch('https://doi.org/' + data.substr(1), {
    		        headers: { 
				Accept: 'application/vnd.citationstyles.csl+json',
				'User-Agent': `mailto:${pluginID}; ${window.navigator.userAgent}`
			}
    	        })
			.then(res => res.json())
			.then(json => span.textContent = _caches[data.substr(1)] = json['is-referenced-by-count'])
			.catch(_ => span.textContent = '');
		span.textContent = 'Loading...';
	} else  // 已获取DOI
		span.textContent = data;
	return span;
}

require('Zotero').ItemTreeManager.registerColumns({
  	dataKey: 'action-citation-counts',
	label: 'Citations',
	dataProvider: item => _caches[item.getField('DOI')] || '#' + item.getField('DOI'),
	pluginID,  // 禁用Tag插件后消失
	renderCell
});
```

**Comment by Edilson-R** [comment]
Please, how can I make the information about the citation count persistent/permanent, and only have it reload if I trigger the event? 

The issue is, I am using Zotero 7, and every time I scroll within a collection, it shows 'loading'. I would like to categorize by citation count, but the way it is now, it's always dynamic

**Reply by volatile-static** [reply]
Well, this is a question. I suggest you don't scroll before they were loaded, and then _switch to another collection and switch back_.
Actually, the citation number were cached once they loaded. The caches would not refresh until the plugin restarted.

**Comment by s1291** [comment]
Many thanks for the script. Below is a screenshot for people who do not understand Chinese:

![image](https://github.com/windingwind/zotero-actions-tags/assets/17838489/18c3605b-7081-4bb2-bb55-edafd1408399)

**Reply by RoboEva** [reply]
Please, what citation is being queried here, Google, Ryan or other?

**Reply by s1291** [reply]
I do not understand your question. But in case you're asking for the source of citation, it's `doi.org`

**Comment by seifer08ms** [comment]
How can use [Semantic Scholar](https://www.semanticscholar.org/) API?  Just like the plugin [ZoteroCitationCountsManager](https://github.com/FrLars21/ZoteroCitationCountsManager)
